### PR TITLE
Add example simulations from hCore to integration test suite

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Adds the popular simulations from hCore's example drop-down menu to the integration test suite to have a quick coverage over a broad range of functionality.

## ⚠️ Known issues

GitHub Actions isn't very fast, maybe we have to lower the experiment values in order to let the tests pass in a reasonable time

## 🐾 Next steps

- Add simulations from https://hash.ai
- Replace Rust behaviors with JavaScript behaviors
- Use the added simulations to gather some timings

## 🔍 What does this change?

- Moves the experiment-module out of the unit-test module in order to allow the usage in other modules as well
- Export projects from hCore and unpack them into the _tests/examples_ directory
- Rename experiments in _tests/examples/<PROJECT>/experiments.json_ to be _snake_case_ in order to use them as function name
- Add *tests/examples/$PROJECT/integration_test.json* and specify the experiments without checking for a specific output (Test will pass if simulation passed successfully)
- Disable unsupported tests (optimization experiments and Rust behaviors - See *Next steps**)
- Add a `--release` flag for running the examples in CI
- A list of exported Projects can be found here: https://github.com/hashintel/hash/pull/250#issuecomment-1027193343

## 👁️ Review guidance

Files changed:
- *.github/workflows/rust.yml* (Added `--release` config)
- *packages/engine/tests/examples/mod.rs* (specify projects)
- *packages/engine/tests/examples/$PROJECT/experiments.json* (names -> *snake_case*)
- *packages/engine/tests/examples/$PROJECT/integration-test.json* (added)
- https://github.com/hashintel/hash/pull/250/commits/ae2653ddb55e7ae9875bcc390652e9507e849dba (Move a few things around to support more integration tests subfolders)

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201707629991362/1201744229736353/f) (_internal_)

## ❓ How to test this?

1. run `cargo build --release`
1. run `cargo test --release examples`

## 📹 Demo

image

### CODE REVIEW

**Comment by TimDiekmann** [comment]
@nonparibus @CiaranMn Could you please take a look if publishing those simulation is fine?
These are the simulations used:
- https://core.hash.ai/@hash/city-infection-model/6.4.2
- https://core.hash.ai/@hash/sugarscape/7.5.0
- https://core.hash.ai/@hash/published-display-behaviors/2.3.0
- https://core.hash.ai/@hash/model-market/4.5.2
- https://core.hash.ai/@hash/wildfires-regrowth/9.8.0
- https://core.hash.ai/@hash/ant-foraging/7.4.0
- https://core.hash.ai/@hash/virus-mutation-and-drug-resistance/3.5.0
- https://core.hash.ai/@hash/rainfall/7.2.2
- https://core.hash.ai/@hash/rumor-mill-public-health-practices/2.2.3
- https://core.hash.ai/@hash/connection-example/1.1.1

Also, I'll add those in a follow up PR:
- https://core.hash.ai/@hash/interconnected-call-center/3.2.1
- https://core.hash.ai/@hash/air-defense-system/1.3.1
- https://core.hash.ai/@hash/city-infection-model-with-vaccine/1.0.2
- https://core.hash.ai/@hash/wholesale-warehouse1/1.1.0

**Comment by CiaranMn** [comment]
> Could you please take a look if publishing those simulation is fine? These are the simulations used (hCore example dropdown)

I'll be honest that I haven't looked through all the source, but if they're pulled from the production examples menu then they are good to use.

One point worth mentioning as it could in theory be the problem of some horribly obscure bug: the Rust behaviors included are not the actual Rust but will be run. Since we do not yet evaluate user-supplied Rust code, but instead use the hardcoded behaviors in `worker/runner/rust`, the code shown in the tests simulations is not necessarily the code that is run. 

Some have diverged, for example the [`age.rs`](https://github.com/hashintel/hash/blob/main/packages/engine/src/worker/runner/rust/behaviors/age.rs) that will be actually run is slightly different (in syntax if not effect) from the age.rs included [here](https://github.com/hashintel/hash/blob/d386720199ce195a7ef0212b86000cd4d714679d/packages/engine/tests/examples/ant_foraging/dependencies/%40hash/age/src/behaviors/age.rs). 

We could potentially replace the Rust behaviors in the examples with symlinks to the actual code that will be run, in case it ever matters. Or maybe we just live with it, since the stacktrace should show which actual `age.rs` was the source of an error, if one occurs.

File: packages/engine/tests/examples/mod.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    // Rust behavior is currently not supported
+    run_test!(city_infection_model, experiment: duration_range_monte_carlo, #[ignore]);
+}
+
+// https://core.hash.ai/@hash/sugarscape/7.5.0
+mod sugarscape {
+    use crate::run_test;
+
+    // optimization experiment is not implemented
+    run_test!(sugarscape, experiment: minimize_the_gini_coefficient, #[ignore]);
```
Comment (Alfred-Mountfield) [inline packages/engine/tests/examples/mod.rs:20]: Lets just not include optimization run_test! calls. 

(Still leave the #[ignore] on the Rust behaviors for now though)

**Review by Alfred-Mountfield — COMMENTED** [review]
I don't see any indication of the versions of the example projects. Can we please include that for each so we have something to compare against when we want to update them

File: packages/engine/tests/examples/mod.rs (hunk 1 above)
Comment (TimDiekmann) [inline packages/engine/tests/examples/mod.rs:20]:
```suggestion
```

File: packages/engine/tests/examples/mod.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    // Rust behavior is currently not supported
+    run_test!(sugarscape, experiment: initial_sugar_linear_sweep, #[ignore]);
+    // Rust behavior is currently not supported
+    run_test!(sugarscape, experiment: agent_density_linspace, #[ignore]);
+    // optimization experiment is not implemented
+    run_test!(sugarscape, experiment: max_avg_sugar, #[ignore]);
```
Comment (TimDiekmann) [inline packages/engine/tests/examples/mod.rs:26]:
```suggestion
```

File: packages/engine/tests/examples/mod.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    // Rust behavior is currently not supported
+    run_test!(model_market, experiment: min_cost_arange, #[ignore]);
+}
+
+// https://core.hash.ai/@hash/wildfires-regrowth/9.8.0
+mod wildfire_regrowth {
+    use crate::run_test;
+
+    // optimization experiment is not implemented
+    run_test!(wildfire_regrowth, experiment: optimal_rates_for_forest_growth, #[ignore]);
```
Comment (TimDiekmann) [inline packages/engine/tests/examples/mod.rs:64]:
```suggestion
```

File: packages/engine/tests/examples/mod.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    // Rust behavior is currently not supported
+    run_test!(wildfire_regrowth, experiment: lightning_chance_linspace, #[ignore]);
+}
+
+// https://core.hash.ai/@hash/ant-foraging/7.4.0
+mod ant_foraging {
+    use crate::run_test;
+
+    // optimization experiment is not implemented
+    run_test!(ant_foraging, experiment: fastest_gathering, #[ignore]);
```
Comment (TimDiekmann) [inline packages/engine/tests/examples/mod.rs:76]:
```suggestion
```

File: packages/engine/tests/examples/mod.rs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+// https://core.hash.ai/@hash/virus-mutation-and-drug-resistance/3.5.0
+mod virus_mutation_and_drug_resistance {
+    use crate::run_test;
+
+    // optimization experiment is not implemented
+    run_test!(virus_mutation_and_drug_resistance, experiment: optimize_vaccine_introduction, #[ignore]);
+    // optimization experiment is not implemented
+    run_test!(virus_mutation_and_drug_resistance, experiment: introduce_vaccine_timestep, #[ignore]);
```
Comment (TimDiekmann) [inline packages/engine/tests/examples/mod.rs:92]:
```suggestion
```

File: packages/engine/tests/examples/mod.rs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    run_test!(virus_mutation_and_drug_resistance, experiment: vaccination_linspace, #[ignore]);
+}
+
+// https://core.hash.ai/@hash/virus-mutation-and-drug-resistance/3.5.0
+mod warehouse_logistics {
+    use crate::run_test;
+
+    run_test!(warehouse_logistics, experiment: alternate_layout);
+    // optimization experiment is not implemented
+    run_test!(warehouse_logistics, experiment: find_optimal_layout, #[ignore]);
```
Comment (TimDiekmann) [inline packages/engine/tests/examples/mod.rs:105]:
```suggestion
```

**Review by TimDiekmann — COMMENTED** [review]
The version is written directly above the module containing the test runs:
```rust
// https://core.hash.ai/@hash/sugarscape/7.5.0
```
The version of sugarscape is 7.5.0

**Comment by Alfred-Mountfield** [comment]
Hmm, I think I'd prefer if it was also explicitly in the actual project folders, but at least we have it written down. Apologies for missing it! I'm happy to approve once the conflicts are resolved

**Review by Alfred-Mountfield — APPROVED** [review]

**Comment by TimDiekmann** [comment]
Removed Label as the bench CI is not needed for merging

**Review by Alfred-Mountfield — APPROVED** [review]

**Pull request merged**